### PR TITLE
Add SourceSeparationBundle to prototype

### PR DESCRIPTION
**Pull request by nateanl**
- Add SourceSeparationBundle class for source separation pipeline
- Add `CONVTASNET_BASE_LIBRI2MIX` that is trained on Libri2Mix dataset.
- Add integration test with example mixture audio and expected scale-invariant signal-to-distortion ratio (Si-SDR) score. The test computes the Si-SDR score with permutation-invariant training (PIT) criterion for all permutations of sources and use the highest value as the final output. The test verifies if the score is equal to or larger than the expected value.

### CODE REVIEW

File: test/integration_tests/conftest.py [hunk 1]
```diff
@@ -40,6 +40,8 @@ def ctc_decoder():
     "fr": "20121212-0900-PLENARY-5-fr_20121212-11_37_04_10.flac",
     "it": "20170516-0900-PLENARY-16-it_20170516-18_56_31_1.flac",
 }
+_MIXTURE_FILENAME = "mix_3729-6852-0037_8463-287645-0000.wav"
+_EXPECTED_FILENAME = "expected_3729-6852-0037_8463-287645-0000.pt"
```
Comment (mthrok) [inline test/integration_tests/conftest.py:44]: This approach is problematic for couple of reasons;
1. The data is not easily recoverable.
2. The internal change of PyTorch can break this test anytime, and it is almost impossible to track down the exact reason, nor we want to spend time on it.

These are the reasons why TTS integration test does not have this kind of assertion.
I recommend to find alternative (not equivalent) way to gauge the correctness of the model.

File: torchaudio/prototype/pipelines/source_separation_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        >>> # Apply source separation on mixture audio.
+        >>> for i, data in enumerate(dataset):
+        >>>     sample_rate, mixture, clean_sources = data
+        >>>     # Make sure the shape of input suits the model requirement.
+        >>>     mixture = mixture.reshape(1, 1, -1)
+        >>>     estimated_sources = separator(mixture)
+        >>>     score = si_snr_pit(estimated_sources, clean_sources) # for demonstration
+        >>>     print(f"Si-SNR score is : {score}.)
+        >>>     break
+        >>> 16.24
```
Comment (skim0514) [inline torchaudio/prototype/pipelines/source_separation_pipeline.py:56]:
```suggestion
        >>> Si-SNR score is : 16.24.
```

File: torchaudio/prototype/pipelines/source_separation_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+class _Separator(torch.nn.Module):
+    def __init__(self, model: torch.nn.Module):
+        super().__init__()
+        self.model = model
+
+    def forward(self, feature: torch.Tensor) -> torch.Tensor:
+        """Separates the input mixture speech into different sources.
+
+        Args:
+            input (torch.Tensor): input tensor with dimensions `[batch, time]`.
```
Comment (skim0514) [inline torchaudio/prototype/pipelines/source_separation_pipeline.py:20]: Is it possible to generalize shape for future inputs that could include things like num_channels

File: torchaudio/prototype/pipelines/source_separation_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+from torchaudio.prototype.models import conv_tasnet_base
+
+
+class _Separator(torch.nn.Module):
+    def __init__(self, model: torch.nn.Module):
+        super().__init__()
+        self.model = model
+
+    def forward(self, feature: torch.Tensor) -> torch.Tensor:
+        """Separates the input mixture speech into different sources.
```
Comment (skim0514) [inline torchaudio/prototype/pipelines/source_separation_pipeline.py:17]:
```suggestion
        """Separates the input mixture audio into different sources.
```

File: test/integration_tests/source_separation_pipeline_test.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        [CONVTASNET_BASE_LIBRI2MIX, "speech_separation"],
+    ],
+)
+def test_source_separation_models(bundle, task, mixture_speech, expected_tensor):
+    """Smoke test of source separation pipeline"""
+    separator = bundle.get_separator()
+    mixture_speech, sample_rate = torchaudio.load(mixture_speech)
+    expected_tensor = torch.load(expected_tensor)
+    mixture_speech = mixture_speech.reshape(1, 1, -1)
+    estimated_sources = separator(mixture_speech)
```
Comment (carolineechen) [inline test/integration_tests/source_separation_pipeline_test.py:19]: would it be sufficient/better to test that `expected_tensor` and `estimated_sources` are equal?

File: test/integration_tests/source_separation_pipeline_test.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    "bundle,task",
+    [
+        [CONVTASNET_BASE_LIBRI2MIX, "speech_separation"],
+    ],
+)
+def test_source_separation_models(bundle, task, mixture_speech, expected_tensor):
+    """Smoke test of source separation pipeline"""
+    separator = bundle.get_separator()
+    mixture_speech, sample_rate = torchaudio.load(mixture_speech)
+    expected_tensor = torch.load(expected_tensor)
```
Comment (mthrok) [inline test/integration_tests/source_separation_pipeline_test.py:17]: Please do not use serialized tensor, rather use regular audio for this

File: test/integration_tests/conftest.py [hunk 7]
```diff
@@ -40,6 +40,12 @@ def ctc_decoder():
     "fr": "20121212-0900-PLENARY-5-fr_20121212-11_37_04_10.flac",
     "it": "20170516-0900-PLENARY-16-it_20170516-18_56_31_1.flac",
 }
+_MIXTURE_FILES = {
+    "speech_separation": "mix_3729-6852-0037_8463-287645-0000.wav",
+}
+_EXPECTED_TENSORS = {
+    "speech_separation": "expected_3729-6852-0037_8463-287645-0000.pt",
```
Comment (mthrok) [inline test/integration_tests/conftest.py:47]: Please replace with this regular audio format.

File: test/integration_tests/source_separation_pipeline_test.py (hunk 5 above)
Comment (mthrok) [reply test/integration_tests/source_separation_pipeline_test.py:19]: @carolineechen It's not the randomness that is an issue, but the fact that future minor changes in PyTorch would slightly change the result from time to time.

Using reference from a past run and comparing against it is not maintainable. It will suddenly break due to some change in core that happens in deep down of some BLAS routine change.

File: test/integration_tests/conftest.py (hunk 7 above)
Comment (nateanl) [reply test/integration_tests/conftest.py:47]: The reason why I used tensor before is the output waveform has large values, after saving it to audio file it causes clipping.

Let me try to save the normalized waveform to see if the peak frequency test can pass.

Comment (mthrok) [reply test/integration_tests/conftest.py:47]: In that case, I think it makes more sense to compare against the ground truth audio, and apply threshold to the SNR between original reference and result.

Comment (nateanl) [reply test/integration_tests/conftest.py:47]: So rather than uploading the estimated waveforms, we can upload the ground truth waveforms and compute the SNR scores then compare it with a given values? Makes sense to me.
Can we use third-party library for metric computation, for example, `mir_eval`?

File: test/integration_tests/source_separation_pipeline_test.py (hunk 5 above)
Comment (nateanl) [reply test/integration_tests/source_separation_pipeline_test.py:19]: hmm actually when I compared the waveforms on CI, the current test already failed, but I guess the error is tolerable if we set up a threshold the test can pass.

File: test/integration_tests/conftest.py [hunk 8]
```diff
@@ -53,6 +62,29 @@ def sample_speech(tmp_path, lang):
     return path
 
 
+@pytest.fixture
+def mixture_source(tmp_path, task):
```
Comment (mthrok) [inline test/integration_tests/conftest.py:66]: Please get rid of `task` parameterization as there is only one at the moment. We can add it when it becomes absolutely necessary.

File: test/integration_tests/conftest.py [hunk 9]
```diff
@@ -53,6 +62,29 @@ def sample_speech(tmp_path, lang):
...
+
+@pytest.fixture
+def clean_sources(tmp_path, task):
+    if task not in _CLEAN_FILES:
+        raise NotImplementedError(f"Unexpected task: {task}")
+    paths = []
+    for file in _CLEAN_FILES[task]:
+        path = tmp_path.parent / file
+        if not path.exists():
+            torchaudio.utils.download_asset(f"test-assets/{file}", path=path)
```
Comment (mthrok) [inline test/integration_tests/conftest.py:83]: Since this is using `torchaudio.utils.download_asset`, `tmp_path` is not necessary. `download_asset` function will resolve the path for you.

File: test/integration_tests/source_separation_pipeline_test.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import torch
+import torchaudio
+from torchaudio.prototype.pipelines import CONVTASNET_BASE_LIBRI2MIX
+
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", "examples"))
+from source_separation.utils.metrics import PIT, sdr
+
+BUNDLE = CONVTASNET_BASE_LIBRI2MIX
+EXPECTED_SCORE = 8.1373
```
Comment (mthrok) [inline test/integration_tests/source_separation_pipeline_test.py:13]: Please put these in test function. they should not be module-level global.

File: torchaudio/prototype/pipelines/source_separation_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    _sample_rate: int
+
+    @property
+    def sample_rate(self) -> int:
+        """Sample rate (in cycles per second) of input waveforms.
+        :type: int
+        """
+        return self._sample_rate
+
+    def get_separator(self) -> torch.nn.Module:
```
Comment (mthrok) [inline torchaudio/prototype/pipelines/source_separation_pipeline.py:53]: Can we call this `get_model`? `separator` does not sound common, and there is no other structure related to this bundle.

File: test/integration_tests/conftest.py [hunk 12]
```diff
@@ -53,6 +58,21 @@ def sample_speech(tmp_path, lang):
     return path
 
 
+@pytest.fixture
+def mixture_source():
+    path = torchaudio.utils.download_asset(f"test-assets/{_MIXTURE_FILE}")
```
Comment (mthrok) [inline test/integration_tests/conftest.py:63]: The use of slash prevents this test from running in Windows. Can you use os.path or pathlib?

File: torchaudio/prototype/pipelines/source_separation_pipeline.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+
+CONVTASNET_BASE_LIBRI2MIX = SourceSeparationBundle(
+    _model_path="models/conv_tasnet_base_libri2mix.pt",
+    _model_factory_func=partial(conv_tasnet_base, num_sources=2),
+    _sample_rate=8000,
+)
+CONVTASNET_BASE_LIBRI2MIX.__doc__ = """Pre-trained ConvTasNet pipeline for source separation.
+    The underlying model is constructed by :py:func:`torchaudio.prototyoe.models.conv_tasnet_base`
+    and utilizes weights trained on Libri2Mix using training script ``lightning_train.py``
```
Comment (mthrok) [inline torchaudio/prototype/pipelines/source_separation_pipeline.py:69]: Can you add citation to Libri2Mix dataset?
